hadoop-yarn-project/hadoop-yarn/hadoop-yarn-client/src/test/java/org/apache/hadoop/yarn/client/TestNoHaRMFailoverProxyProvider.java

```diff
@@ -0,0 +1,269 @@
+import org.junit.Test;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.Proxy;
+import java.net.InetSocketAddress;
+import java.util.List;
+
+import static org.mockito.Mockito.any;
+import static org.mockito.Mockito.eq;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+public class TestNoHaRMFailoverProxyProvider {
+    private final int NODE_MANAGER_COUNT = 1;
+    private Configuration conf;
+
+    @Before
+    public void setUp() throws IOException, YarnException {
+        conf = new YarnConfiguration();
+    }
+
+    @Test
+    public void testRestartedRM() throws Exception {
+        try {
+            MiniYARNCluster cluster =
+                new MiniYARNCluster(
+                    "testRestartedRMNegative", NODE_MANAGER_COUNT, 1, 1);
+            YarnClient rmClient = YarnClient.createYarnClient();
+            cluster.init(conf);
+            cluster.start();
+            final Configuration yarnConf = cluster.getConfig();
+            rmClient = YarnClient.createYarnClient();
+            rmClient.init(yarnConf);
+            rmClient.start();
+            List<NodeReport> nodeReports = rmClient.getNodeReports();
+            Assert.assertEquals(
+                    "The proxy didn't get expected number of node reports",
+                    NODE_MANAGER_COUNT, nodeReports.size());
+        } finally {
+            if (rmClient != null) {
+                rmClient.stop();
+            }
+            cluster.stop();
+        }
+    }
+
+    /**
+     * Tests the proxy generated by {@link AutoRefreshNoHARMFailoverProxyProvider}
+     * will connect to RM.
+     */
+    @Test
+    public void testConnectingToRM() throws Exception {
+        conf.setClass(YarnConfiguration.CLIENT_FAILOVER_NO_HA_PROXY_PROVIDER,
+                AutoRefreshNoHARMFailoverProxyProvider.class, RMFailoverProxyProvider.class);
+
+        try {
+            MiniYARNCluster cluster =
+                new MiniYARNCluster(
+                    "testRestartedRMNegative", NODE_MANAGER_COUNT, 1, 1);
+            YarnClient rmClient = null;            cluster.init(conf);
+            cluster.start();
+            final Configuration yarnConf = cluster.getConfig();
+            rmClient = YarnClient.createYarnClient();
+            rmClient.init(yarnConf);
+            rmClient.start();
+            List<NodeReport> nodeReports = rmClient.getNodeReports();
+            Assert.assertEquals(
+                    "The proxy didn't get expected number of node reports",
+                    NODE_MANAGER_COUNT, nodeReports.size());
+        } finally {
+            if (rmClient != null) {
+                rmClient.stop();
+            }
+            cluster.stop();
+        }
+    }
+
+    @Test
+    public void testDefaultFPPGetOneProxy() throws Exception {
```

Comment:
It would be good to have javadocs for each test to describe the high level goals.